From eab365c6aa837f128026fb7bf139bf64b50d48f3 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Mon, 2 Feb 2026 14:41:58 +0400
Subject: [PATCH] feat: gibs: damage-based random velocity

Note that this changes the network protocol.
But it's still compatible with vanilla.
See new comment in `g_combat.c`.

We had to move the code that resets
`damage_blood`, `damage_armor`, `damage_knockback`
because otherwise they would stop getting reset
when the player dies,
because `P_DamageFeedback` early-returns when `PM_DEAD`.
---
 code/cgame/cg_effects.c    |  7 +++--
 code/cgame/cg_event.c      |  9 ++++--
 code/cgame/cg_local.h      |  7 ++++-
 code/cgame/cg_localents.c  |  5 ++-
 code/cgame/cg_main.c       |  5 ++-
 code/cgame/cg_servercmds.c | 12 +++++++
 code/game/bg_public.h      |  7 +++++
 code/game/g_active.c       | 14 ++++-----
 code/game/g_combat.c       | 64 +++++++++++++++++++++++++++++++-------
 code/game/g_local.h        |  1 +
 code/game/g_main.c         |  2 ++
 11 files changed, 108 insertions(+), 25 deletions(-)

diff --git a/code/cgame/cg_effects.c b/code/cgame/cg_effects.c
index 65d5450a..f6d21b38 100644
--- a/code/cgame/cg_effects.c
+++ b/code/cgame/cg_effects.c
@@ -635,7 +635,7 @@ Generated a bunch of gibs launching out from the bodies location
 #define	GIB_VELOCITY		250
 #define	GIB_JUMP			250
 void CG_GibPlayer( const vec3_t playerOrigin, const vec3_t playerAngles,
-					const vec3_t playerVelocity,
+					const vec3_t playerVelocity, const int knockbackSpeed,
 					const lerpFrame_t *bodyAnimation ) {
 	vec3_t	baseOrigin, origin, velocity;
 	// Generally only the head should have pitch,
@@ -649,7 +649,10 @@ void CG_GibPlayer( const vec3_t playerOrigin, const vec3_t playerAngles,
 	// to account for crounching.
 	float playerHeight = 32 - MINS_Z;
 	float playerRadius = PLAYER_RADIUS;
-	float baseRandomVelocity = cg_gibsExtraRandomVelocity.value;
+	float baseRandomVelocity =
+		cg_gibsExtraRandomVelocity.value +
+		cg_gibsRandomVelocityFromKnockback.value * knockbackSpeed;
+
 	vec3_t playerVelocityScaled;
 	float jump = cg_gibsExtraVerticalVelocity.value;
 	int numGibs = cg_gibs.value * DEFAULT_NUM_GIBS;
diff --git a/code/cgame/cg_event.c b/code/cgame/cg_event.c
index 065175f7..d2986002 100644
--- a/code/cgame/cg_event.c
+++ b/code/cgame/cg_event.c
@@ -1222,6 +1222,11 @@ void CG_EntityEvent( centity_t *cent, vec3_t position ) {
 		if (cg_oldGibs.integer) {
 			CG_GibPlayerOld( cent->lerpOrigin );
 		} else {
+			int knockbackSpeed = cgs.g_gibsNewEvGibPlayerParmProtocol == 1
+				? es->eventParm * COMBAT_EV_GIB_PLAYER_ARG_DIVISOR
+				// Just use the default knockback speed for 100 damage.
+				: 100 * 1000 / COMBAT_PLAYER_MASS;
+
 			if ( es->number == cg.snap->ps.clientNum ) {
 				// Apparently at this point `es->pos.trDelta` doesn't yet have
 				// the knockback from the damage that gibbed us,
@@ -1230,11 +1235,11 @@ void CG_EntityEvent( centity_t *cent, vec3_t position ) {
 				// if it's ourself.
 				// `cent->pe.torso` also appears to be not good here.
 				CG_GibPlayer( cent->lerpOrigin, cent->lerpAngles,
-					cg.predictedPlayerState.velocity,
+					cg.predictedPlayerState.velocity, knockbackSpeed,
 					&cg.predictedPlayerEntity.pe.torso );
 			} else {
 				CG_GibPlayer( cent->lerpOrigin, cent->lerpAngles,
-					es->pos.trDelta,
+					es->pos.trDelta, knockbackSpeed,
 					&cent->pe.torso );
 			}
 		}
diff --git a/code/cgame/cg_local.h b/code/cgame/cg_local.h
index f1bae3d0..0e07221f 100644
--- a/code/cgame/cg_local.h
+++ b/code/cgame/cg_local.h
@@ -1014,6 +1014,9 @@ typedef struct {
 	char			redTeam[MAX_QPATH];
 	char			blueTeam[MAX_QPATH];
 
+	// parsed from systeminfo
+	int				g_gibsNewEvGibPlayerParmProtocol;
+
 	int				voteTime;
 	int				voteYes;
 	int				voteNo;
@@ -1096,6 +1099,7 @@ extern	vmCvar_t		cg_gibs;
 extern	vmCvar_t		cg_oldGibs;
 extern	vmCvar_t		cg_gibsInheritPlayerVelocity;
 extern	vmCvar_t		cg_gibsExtraRandomVelocity;
+extern	vmCvar_t		cg_gibsRandomVelocityFromKnockback;
 extern	vmCvar_t		cg_gibsExtraVerticalVelocity;
 extern	vmCvar_t		cg_gibsBounceFactor;
 extern	vmCvar_t		cg_gibsRotationFactor;
@@ -1426,7 +1430,7 @@ void CG_LightningBoltBeam( vec3_t start, vec3_t end );
 void CG_ScorePlum( int client, vec3_t org, int score );
 
 void CG_GibPlayer( const vec3_t playerOrigin, const vec3_t playerAngles,
-				const vec3_t playerVelocity,
+				const vec3_t playerVelocity, const int knockbackSpeed,
 				const lerpFrame_t *bodyAnimation );
 void CG_GibPlayerOld( vec3_t playerOrigin );
 void CG_BigExplode( vec3_t playerOrigin );
@@ -1467,6 +1471,7 @@ void CG_InitConsoleCommands( void );
 //
 void CG_ExecuteNewServerCommands( int latestSequence );
 void CG_ParseServerinfo( void );
+void CG_ParseSysteminfo( void );
 void CG_SetConfigValues( void );
 void CG_ShaderStateChanged(void);
 #ifdef MISSIONPACK
diff --git a/code/cgame/cg_localents.c b/code/cgame/cg_localents.c
index 1c1f214e..f1ef96cf 100644
--- a/code/cgame/cg_localents.c
+++ b/code/cgame/cg_localents.c
@@ -701,11 +701,14 @@ void CG_AddInvulnerabilityJuiced( localEntity_t *le ) {
 	}
 	if ( t > 5000 ) {
 		vec3_t angles;
+		// Just use the default knockback speed for 200 damage.
+		int knockbackSpeed = 200 * 1000 / COMBAT_PLAYER_MASS;
 		// Angles don't matter much here.
 		VectorClear( angles );
 
 		le->endTime = 0;
-		CG_GibPlayer( le->refEntity.origin, angles, le->pos.trDelta, NULL );
+		CG_GibPlayer( le->refEntity.origin, angles, le->pos.trDelta,
+			knockbackSpeed, NULL );
 	}
 	else {
 		trap_R_AddRefEntityToScene( &le->refEntity );
diff --git a/code/cgame/cg_main.c b/code/cgame/cg_main.c
index 38d7b079..e8816489 100644
--- a/code/cgame/cg_main.c
+++ b/code/cgame/cg_main.c
@@ -102,6 +102,7 @@ vmCvar_t	cg_gibs;
 vmCvar_t	cg_oldGibs;
 vmCvar_t	cg_gibsInheritPlayerVelocity;
 vmCvar_t	cg_gibsExtraRandomVelocity;
+vmCvar_t	cg_gibsRandomVelocityFromKnockback;
 vmCvar_t	cg_gibsExtraVerticalVelocity;
 vmCvar_t	cg_gibsBounceFactor;
 vmCvar_t	cg_gibsRotationFactor;
@@ -226,7 +227,8 @@ static cvarTable_t cvarTable[] = {
 	{ &cg_gibs, "cg_gibs", "1.0", CVAR_ARCHIVE  },
 	{ &cg_oldGibs, "cg_oldGibs", "0", CVAR_ARCHIVE  },
 	{ &cg_gibsInheritPlayerVelocity, "cg_gibsInheritPlayerVelocity", "1.0", CVAR_ARCHIVE  },
-	{ &cg_gibsExtraRandomVelocity, "cg_gibsExtraRandomVelocity", "250", CVAR_ARCHIVE  },
+	{ &cg_gibsExtraRandomVelocity, "cg_gibsExtraRandomVelocity", "175", CVAR_ARCHIVE  },
+	{ &cg_gibsRandomVelocityFromKnockback, "cg_gibsRandomVelocityFromKnockback", "0.15", CVAR_ARCHIVE  },
 	{ &cg_gibsExtraVerticalVelocity, "cg_gibsExtraVerticalVelocity", "100", CVAR_ARCHIVE  },
 	{ &cg_gibsBounceFactor, "cg_gibsBounceFactor", "0.4", CVAR_ARCHIVE  },
 	{ &cg_gibsRotationFactor, "cg_gibsRotationFactor", "1.0", CVAR_ARCHIVE  },
@@ -1910,6 +1912,7 @@ void CG_Init( int serverMessageNum, int serverCommandSequence, int clientNum ) {
 	cgs.levelStartTime = atoi( s );
 
 	CG_ParseServerinfo();
+	CG_ParseSysteminfo();
 
 	// load the new map
 	CG_LoadingString( "collision map" );
diff --git a/code/cgame/cg_servercmds.c b/code/cgame/cg_servercmds.c
index 5829e8f3..fe13adec 100644
--- a/code/cgame/cg_servercmds.c
+++ b/code/cgame/cg_servercmds.c
@@ -173,6 +173,16 @@ void CG_ParseServerinfo( void ) {
 	trap_Cvar_Set("g_blueTeam", cgs.blueTeam);
 }
 
+
+void CG_ParseSysteminfo( void ) {
+	const char	*info;
+
+	info = CG_ConfigString( CS_SYSTEMINFO );
+
+	cgs.g_gibsNewEvGibPlayerParmProtocol =
+		atoi( Info_ValueForKey( info, "g_gibsNewEvGibPlayerParmProtocol" ) );
+}
+
 /*
 ==================
 CG_ParseWarmup
@@ -302,6 +312,8 @@ static void CG_ConfigStringModified( void ) {
 	// do something with it if necessary
 	if ( num == CS_MUSIC ) {
 		CG_StartMusic();
+	} else if ( num == CS_SYSTEMINFO ) {
+		CG_ParseSysteminfo();
 	} else if ( num == CS_SERVERINFO ) {
 		CG_ParseServerinfo();
 	} else if ( num == CS_WARMUP ) {
diff --git a/code/game/bg_public.h b/code/game/bg_public.h
index 7d203fb0..07bb9af2 100644
--- a/code/game/bg_public.h
+++ b/code/game/bg_public.h
@@ -202,6 +202,13 @@ void Pmove (pmove_t *pmove);
 
 //===================================================================================
 
+#define COMBAT_PLAYER_MASS 200
+// A divisor of knockback speed, to fit it into one byte.
+// By dividing by 8 we can represent a speed of up to (255 * 8) = 2040.
+// For comparison, with `g_knockback` of 1000 and `MAX_KNOCKBACK` of 200
+// the max knockback speed in most situations is 1000.
+#define COMBAT_EV_GIB_PLAYER_ARG_DIVISOR 8
+
 
 // player_state->stats[] indexes
 // NOTE: may not have more than 16
diff --git a/code/game/g_active.c b/code/game/g_active.c
index c8d8dbe5..b27ddbd0 100644
--- a/code/game/g_active.c
+++ b/code/game/g_active.c
@@ -78,13 +78,6 @@ void P_DamageFeedback( gentity_t *player ) {
 
 
 	client->ps.damageCount = count;
-
-	//
-	// clear totals
-	//
-	client->damage_blood = 0;
-	client->damage_armor = 0;
-	client->damage_knockback = 0;
 }
 
 
@@ -1189,6 +1182,13 @@ void ClientEndFrame( gentity_t *ent ) {
 	}
 	SendPendingPredictableEvents( &ent->client->ps );
 
+	//
+	// clear damage totals
+	//
+	ent->client->damage_blood = 0;
+	ent->client->damage_armor = 0;
+	ent->client->damage_knockback = 0;
+
 	// set the bit for the reachability area the client is currently in
 //	i = trap_AAS_PointReachabilityAreaIndex( ent->client->ps.origin );
 //	ent->client->areabits[i >> 3] |= 1 << (i & 7);
diff --git a/code/game/g_combat.c b/code/game/g_combat.c
index 30c404fe..ea324618 100644
--- a/code/game/g_combat.c
+++ b/code/game/g_combat.c
@@ -229,17 +229,14 @@ void LookAtKiller( gentity_t *self, gentity_t *inflictor, gentity_t *attacker )
 	self->client->ps.stats[STAT_DEAD_YAW] = vectoyaw ( dir );
 }
 
+#define MAX_KNOCKBACK 200
 /*
 ==================
 KnockbackToKnockbackSpeed
 ==================
 */
 static float KnockbackToKnockbackSpeed( int knockback ) {
-	float	mass;
-
-	mass = 200;
-
-	return g_knockback.value * (float)knockback / mass;
+	return g_knockback.value * (float)knockback / COMBAT_PLAYER_MASS;
 }
 
 
@@ -248,7 +245,8 @@ static float KnockbackToKnockbackSpeed( int knockback ) {
 GibEntity
 ==================
 */
-void GibEntity( gentity_t *self, int killer ) {
+void GibEntity( gentity_t *self, int killer, int damageBloodFallback ) {
+	int eventParm = killer;
 	gentity_t *ent;
 	int i;
 
@@ -267,7 +265,51 @@ void GibEntity( gentity_t *self, int killer ) {
 			break;
 		}
 	}
-	G_AddEvent( self, EV_GIB_PLAYER, killer );
+
+	// In vanilla Quake the meaning of the `EV_GIB_PLAYER` eventParm
+	// is `killer`.
+	// But it is unused client-side, so it's safe for us to change its meaning
+	// (i.e. to change the network protocol).
+	// However, some mods might in fact rely on it,
+	// so let's have a CVAR to keep the old behavior.
+	//
+	// Note that we're not checking `g_oldGibs`, because in itself
+	// this does not affect behavior:
+	// we're simply providing the client with the knockback info,
+	// and whether to use that into is up to `cg_oldGibs`.
+	if ( g_gibsNewEvGibPlayerParmProtocol.integer == 1 ) {
+		int damage;
+		float knockbackSpeed;
+
+		// We prefer actual damage over `client->damage_knockback`
+		// because `damage_knockback` is sometimes undesirably 0. Namely:
+		// - when the target is a dead body, with `FL_NO_KNOCKBACK`.
+		// - when the knockback `dir` is not provided to `G_Damage`,
+		//   such as with crushers.
+		//
+		// Most of the time (but not always e.g. with lava)
+		// "no knockback" means "the player should not be moved
+		// in any particular direction",
+		// and not that "their gibs should stay put".
+		damage = self->client
+			? self->client->damage_blood + self->client->damage_armor
+			: damageBloodFallback;
+		if ( damage > MAX_KNOCKBACK ) {
+			damage = MAX_KNOCKBACK;
+		}
+
+		knockbackSpeed = KnockbackToKnockbackSpeed( damage );
+
+		// Fit it into one byte.
+		eventParm = knockbackSpeed / COMBAT_EV_GIB_PLAYER_ARG_DIVISOR;
+		if (eventParm > 255) {
+			eventParm = 255;
+		}
+	} else {
+		eventParm = killer;
+	}
+	G_AddEvent( self, EV_GIB_PLAYER, eventParm );
+
 	self->takedamage = qfalse;
 	self->s.eType = ET_INVISIBLE;
 	self->r.contents = 0;
@@ -287,7 +329,7 @@ void body_die( gentity_t *self, gentity_t *inflictor, gentity_t *attacker, int d
 		return;
 	}
 
-	GibEntity( self, 0 );
+	GibEntity( self, 0, damage );
 }
 
 
@@ -646,7 +688,7 @@ void player_die( gentity_t *self, gentity_t *inflictor, gentity_t *attacker, int
 
 	if ( (self->health <= GIB_HEALTH && !(contents & CONTENTS_NODROP) && g_blood.integer) || meansOfDeath == MOD_SUICIDE) {
 		// gib death
-		GibEntity( self, killer );
+		GibEntity( self, killer, damage );
 	} else {
 		// normal death
 		static int lastDeath;
@@ -1003,8 +1045,8 @@ void G_Damage( gentity_t *targ, gentity_t *inflictor, gentity_t *attacker,
 	}
 
 	knockback = damage;
-	if ( knockback > 200 ) {
-		knockback = 200;
+	if ( knockback > MAX_KNOCKBACK ) {
+		knockback = MAX_KNOCKBACK;
 	}
 	if ( targ->flags & FL_NO_KNOCKBACK ) {
 		knockback = 0;
diff --git a/code/game/g_local.h b/code/game/g_local.h
index c7f7d971..25cca91d 100644
--- a/code/game/g_local.h
+++ b/code/game/g_local.h
@@ -726,6 +726,7 @@ extern	vmCvar_t	g_doWarmup;
 extern	vmCvar_t	g_blood;
 extern	vmCvar_t	g_oldGibs;
 extern	vmCvar_t	g_gibsMissileDirectionKnockbackWeight;
+extern	vmCvar_t	g_gibsNewEvGibPlayerParmProtocol;
 extern	vmCvar_t	g_allowVote;
 extern	vmCvar_t	g_teamAutoJoin;
 extern	vmCvar_t	g_teamForceBalance;
diff --git a/code/game/g_main.c b/code/game/g_main.c
index ca1e8fe7..05470792 100644
--- a/code/game/g_main.c
+++ b/code/game/g_main.c
@@ -71,6 +71,7 @@ vmCvar_t	g_logfileSync;
 vmCvar_t	g_blood;
 vmCvar_t	g_oldGibs;
 vmCvar_t	g_gibsMissileDirectionKnockbackWeight;
+vmCvar_t	g_gibsNewEvGibPlayerParmProtocol;
 vmCvar_t	g_podiumDist;
 vmCvar_t	g_podiumDrop;
 vmCvar_t	g_allowVote;
@@ -166,6 +167,7 @@ static cvarTable_t		gameCvarTable[] = {
 	// Note that this affects not just the gibs
 	// but also the camera velocity of the gibbed player.
 	{ &g_gibsMissileDirectionKnockbackWeight, "g_gibsMissileDirectionKnockbackWeight", "0.5", CVAR_ARCHIVE, 0, qfalse },
+	{ &g_gibsNewEvGibPlayerParmProtocol, "g_gibsNewEvGibPlayerParmProtocol", "1", CVAR_SYSTEMINFO | CVAR_ARCHIVE, 0, qfalse },
 
 	{ &g_podiumDist, "g_podiumDist", "80", 0, 0, qfalse },
 	{ &g_podiumDrop, "g_podiumDrop", "70", 0, 0, qfalse },
-- 
2.43.0

